README
NAME
    Web::Util::ExtPaging - Paginate DBIx::Class::ResultSets for ExtJS
    consumption

VERSION
    version 0.001002

SYNOPSIS
      package MyApp::People;

      use Web::Simple;
      use JSON::MaybeXS;
      use Web::Util::ExtPaging;

      sub dispatch_request {
        my $people_rs = get_rs();

        sub (/people) {
          [
             200,
             [ 'Content-type', 'application/json' ],
             [ encode_json(ext_paginate($rs->search(undef, { rows => 25 }))) ],
          ]
        },
        sub (/people_lite) {
          [
             200,
             [ 'Content-type', 'application/json' ],
             [
                encode_json(ext_paginate(
                   $rs->search(undef, { rows => 25 }), sub {
                      my $person = shift;
                      return {
                         first_name => $person->first_name,
                         last_name => $person->last_name,
                      }
                   },
                ))
             ],
          ]
        },
        sub (/people_more_different) {
          [
             200,
             [ 'Content-type', 'application/json' ],
             [
                # this will call the 'foo' method on each person and put the
                # returned value into the datastructure
                encode_json(ext_paginate(
                   $rs->search(undef, { rows => 25 }), 'foo',
                ))
             ],
          ]
        },
        sub (/programmers_do_it_by_hand) {
          [
             200,
             [ 'Content-type', 'application/json' ],
             [ encode_json(ext_parcel([qw( foo bar baz )], 10)) ],
          ]
        },
        sub (/programmers_do_it_by_hand_partially) {
          [
             200,
             [ 'Content-type', 'application/json' ],
             # defaults total to amount of items passed in
             [ encode_json(ext_parcel([qw( foo bar baz )])) ],
          ]
        },
        sub () { [ 404, [ 'Content-type', 'text/plain' ], [ 'not found' ] ] }
      }

DESCRIPTION
    This module is mostly for sending DBIx::Class paginated data to ExtJS
    based javascript code.

METHODS
  ext_paginate
      my $json      = ext_paginate($resultset, { root => 'root' });
      my $json_str  = json_encode($json);

   Description
    Returns a structure like the following from the ResultSet:

      {
         data  => \@results,
         total => $count_before_pagination
      }

   Valid arguments are:
      rs - paginated ResultSet to get the data from
      (optional) coderef - any valid scalar that can be called on the result object
      (optional) config - passed to ext_parcel

  ext_parcel
      my $items    = [qw{foo bar baz}];
      my $total    = 7;
      my $json     = $self->ext_parcel($data, $total, { root => 'root' });
      my $json_str = to_json($json);

   Description
    Returns a structure like the following:

      {
         data  => [@{$items}],
         total => $total || scalar @{$items}
      }

   Valid arguments are:
      list  - a list of anything you want to be in the data structure
      total - whatever you want to say the total is.  Defaults to size of
              the list passed in.
      (optional) config - a hashref containing root or total_property.  root is the
              key used to store the data under, total_property is the key used to
              store the total under

SEE ALSO
    Catalyst::TraitFor::Controller::DoesExtPaging, which this module was
    factored out of.

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Web-Util-ExtPaging

0.001002  2014-11-03 16:21:05-06:00 America/Chicago
 - Transparently support resultsets using HRI (Wes Malone)

0.001001  2013-05-31 23:00:29 CST6CDT
 - Remove unneeded dep on parent.pm

0.001000  2013-05-25 13:22:30 CST6CDT
 - Initial Release, factored out of
   Catalyst::TraitFor::Controller::DoesExtPaging



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Paginate DBIx::Class::ResultSets for ExtJS consumption",
  "AUTHOR" => "Arthur Axel \"fREW\" Schmidt <frioux+cpan\@gmail.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Web-Util-ExtPaging",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Web::Util::ExtPaging",
  "PREREQ_PM" => {
    "Sub::Exporter::Progressive" => "0.00101"
  },
  "TEST_REQUIRES" => {
    "DBIx::Class" => "0.0825",
    "SQL::Translator" => "0.11016",
    "Test::Deep" => "0.11",
    "Test::More" => "0.98"
  },
  "VERSION" => "0.001002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBIx::Class" => "0.0825",
  "SQL::Translator" => "0.11016",
  "Sub::Exporter::Progressive" => "0.00101",
  "Test::Deep" => "0.11",
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Web/Util/ExtPaging.pm
package Web::Util::ExtPaging;
$Web::Util::ExtPaging::VERSION = '0.001002';
# ABSTRACT: Paginate DBIx::Class::ResultSets for ExtJS consumption

use strict;
use warnings;

use Sub::Exporter::Progressive -setup => {
  exports => [qw( ext_paginate ext_parcel )],
  groups => {
    default => [qw( ext_paginate ext_parcel )],
  },
};

sub ext_paginate {
   my $resultset = shift;
   my $method    = shift || 'TO_JSON';
   my $config    = shift;

   if (ref $method && ref $method ne 'CODE') {
      $config = $method;
      $method = 'TO_JSON';
   }

   return ext_parcel(
      $resultset->result_class->isa('DBIx::Class::ResultClass::HashRefInflator') ?
         [$resultset->all] :
         [map $_->$method, $resultset->all],
      $resultset->is_paged
         ? ($resultset->pager->total_entries)
         : (),
      ($config ? $config : () ),
   );
}

sub ext_parcel {
   my ($values, $total, $config) = @_;

   if (ref $total) {
      $config = $total;
      $total = undef;
   }

   $total  ||= @$values;
   $config ||= {};

   return {
      ($config->{root}||'data') => $values,
      ($config->{total_property}||'total') => $total,
   };
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Web::Util::ExtPaging - Paginate DBIx::Class::ResultSets for ExtJS consumption

=head1 VERSION

version 0.001002

=head1 SYNOPSIS

  package MyApp::People;

  use Web::Simple;
  use JSON::MaybeXS;
  use Web::Util::ExtPaging;

  sub dispatch_request {
    my $people_rs = get_rs();

    sub (/people) {
      [
         200,
         [ 'Content-type', 'application/json' ],
         [ encode_json(ext_paginate($rs->search(undef, { rows => 25 }))) ],
      ]
    },
    sub (/people_lite) {
      [
         200,
         [ 'Content-type', 'application/json' ],
         [
            encode_json(ext_paginate(
               $rs->search(undef, { rows => 25 }), sub {
                  my $person = shift;
                  return {
                     first_name => $person->first_name,
                     last_name => $person->last_name,
                  }
               },
            ))
         ],
      ]
    },
    sub (/people_more_different) {
      [
         200,
         [ 'Content-type', 'application/json' ],
         [
            # this will call the 'foo' method on each person and put the
            # returned value into the datastructure
            encode_json(ext_paginate(
               $rs->search(undef, { rows => 25 }), 'foo',
            ))
         ],
      ]
    },
    sub (/programmers_do_it_by_hand) {
      [
         200,
         [ 'Content-type', 'application/json' ],
         [ encode_json(ext_parcel([qw( foo bar baz )], 10)) ],
      ]
    },
    sub (/programmers_do_it_by_hand_partially) {
      [
         200,
         [ 'Content-type', 'application/json' ],
         # defaults total to amount of items passed in
         [ encode_json(ext_parcel([qw( foo bar baz )])) ],
      ]
    },
    sub () { [ 404, [ 'Content-type', 'text/plain' ], [ 'not found' ] ] }
  }

=head1 DESCRIPTION

This module is mostly for sending L<DBIx::Class> paginated data to ExtJS based javascript code.

=head1 METHODS

=head2 ext_paginate

  my $json      = ext_paginate($resultset, { root => 'root' });
  my $json_str  = json_encode($json);

=head3 Description

Returns a structure like the following from the ResultSet:

  {
     data  => \@results,
     total => $count_before_pagination
  }

=head3 Valid arguments are:

  rs - paginated ResultSet to get the data from
  (optional) coderef - any valid scalar that can be called on the result object
  (optional) config - passed to ext_parcel

=head2 ext_parcel

  my $items    = [qw{foo bar baz}];
  my $total    = 7;
  my $json     = $self->ext_parcel($data, $total, { root => 'root' });
  my $json_str = to_json($json);

=head3 Description

Returns a structure like the following:

  {
     data  => [@{$items}],
     total => $total || scalar @{$items}
  }

=head3 Valid arguments are:

  list  - a list of anything you want to be in the data structure
  total - whatever you want to say the total is.  Defaults to size of
          the list passed in.
  (optional) config - a hashref containing root or total_property.  root is the
          key used to store the data under, total_property is the key used to
          store the total under

=head1 SEE ALSO

L<Catalyst::TraitFor::Controller::DoesExtPaging>, which this module was factored
out of.

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
